File: bench/benches/deforest.go
package main

// sum of squares of the odd numbers in [0, N) -- a map/filter/fold pipeline.
// the idiomatic Go chain materializes intermediate slices each call; the
// per-element lambda overhead is the point. checksum = 1333333330000.
func main() {
	bench("deforest", func() int64 {
		odd := func(x int64) bool { return x%2 == 1 }
		sq := func(x int64) int64 { return x * x }
		var s int64
		filtered := make([]int64, 0, 20000)
		for i := int64(0); i < 20000; i++ {
			if odd(i) {
				filtered = append(filtered, i)
			}
		}
		mapped := make([]int64, len(filtered))
		for i, v := range filtered {
			mapped[i] = sq(v)
		}
		for _, v := range mapped {
			s += v
		}
		return s
	})
}

File: bench/benches/deforest.hs
{-# LANGUAGE BangPatterns #-}
import Bench
import Data.List (foldl')

-- sum of squares of the odd numbers in [0, N) -- a map/filter/fold pipeline.
-- checksum = 1333333330000. The index source is pinned through `with` (the rep
-- loop's argument) so GHC runs the real O(n) pipeline each rep rather than fold
-- it to a closed form or share it as a CAF.
n :: Int
n = 20000

work :: [Int] -> Integer
work xs = foldl' (\ !a x -> a + x) 0
            [ sq | x <- xs, odd x
                 , let sq = fromIntegral x * fromIntegral x :: Integer ]

main :: IO ()
main = bench "deforest" (\u -> work (with u [0 .. n - 1]))
